godot-ffi/src/toolbox.rs

```diff
@@ -329,3 +307,94 @@ pub(crate) fn read_version_string(version_ptr: &sys::GDExtensionGodotVersion) ->
 }
 
 const INFO: &str = "\nMake sure gdext and Godot are compatible: https://godot-rust.github.io/book/gdext/advanced/compatibility.html";
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+// Private abstractions
+// Dont use abstractions made here outside this crate, if needed then we should discuss making it more of a first-class
+// abstraction like `godot-cell`.
+
+/// Module to encapsulate `UnsafeOnceCell`.
+mod unsafe_once_cell {
+    use std::{cell::UnsafeCell, hint::unreachable_unchecked};
+
+    /// A cell which can only be written to once, where the caller is responsible for synchronization.
+    ///
+    /// Similar to a [`OnceLock`](std::sync::OnceLock), but without the overhead of locking for initialization. In most cases the compiler
+    /// seems able to optimize `OnceLock` to equivalent code. But this guaranteed does not have any overhead during
+    pub(crate) struct UnsafeOnceCell<T> {
+        // Invariant: Is `None` until initialized, and then never modified after (except, possibly, through interior mutability).
+        cell: UnsafeCell<Option<T>>,
+    }
+
+    impl<T> UnsafeOnceCell<T> {
+        /// Creates a new empty cell.
+        pub const fn new() -> Self {
+            Self {
+                cell: UnsafeCell::new(None),
+            }
+        }
+
+        /// Initialize the value stored in this cell.
+        ///
+        /// # Safety
+        ///
+        /// - Must only be called once.
+        /// - Calls to this method must not happen concurrently with a call to any other method on this cell.
+        // Note that the other methods of this cell do not have a safety invariant that they are not called concurrently with `set`.
+        // This is because doing so would violate the safety invariants of `set` and so they do not need to explicitly have that as a
+        // safety invariant as well. This has the added benefit that `is_initialized` can be a safe method.
```

Comment:
Should this be elevated to Rustdoc, after an empty line?

It probably doesn't matter much in practice because this is not public docs, but if someone runs `cargo doc` including private items, this could be helpful to read as well.

Reply:
I guess yeah, i figured that in practice when using it you would not really need to know this as when you try to uphold the safety invariants of `set` you end up just doing this anyway. So it's more an explanation for people trying to understand the code as written instead. But i mean it doesn't really hurt?